File: javascript/phone-number/phone-number.js
var PhoneNumber = function(phone) {
  var clean = function() {
    var scrub = phone.replace(/[^\d]/g, '');
    if (scrub.length == 11 ) { scrub = scrub.replace(/^1/,'') }
    if (scrub.length == 10) return scrub;
    return '0000000000';
  }();

  var parts = clean.match(/^(.{3})(.{3})(.{4})$/);

  this.number = function() {
    return clean;
  };

  this.areaCode = function() {
    return parts[1];
  };

  this.toString = function() {
    return '(' + parts[1] + ') ' + parts[2] + '-' + parts[3];
  };
};

module.exports = PhoneNumber;

File: ruby/phone-number/phone_number.rb
class PhoneNumber
  def initialize(string)
    @number = string.gsub(/[^\d]/,'')
    @number = "0000000000" unless (string =~ /[a-zA-Z]/).nil?
    @number = "0000000000" if number.to_i > 19_999_999_999
    @number = "0000000000" if number.to_i <= 999_999_999
    @number = number.gsub(/1(\d{10})/) { $1 }
  end

  attr_reader :number

  def area_code
    number[0..2]
  end

  def to_s
    "(#{area_code}) #{number[3..5]}-#{number[6..9]}"
  end

end
